x-pack/plugin/searchable-snapshots/qa/s3/build.gradle

```diff
@@ -0,0 +1,82 @@
+import org.elasticsearch.gradle.info.BuildParams
+
+import static org.elasticsearch.gradle.PropertyNormalization.IGNORE_VALUE
+
+apply plugin: 'elasticsearch.standalone-rest-test'
+apply plugin: 'elasticsearch.rest-test'
+apply plugin: 'elasticsearch.test.fixtures'
+
+final Project fixture = project(':test:fixtures:s3-fixture')
+final Project repositoryPlugin = project(':plugins:repository-s3')
+
+dependencies {
+  testCompile project(path: xpackModule('searchable-snapshots'), configuration: 'testArtifacts')
+  testCompile repositoryPlugin
+  testCompile fixture
+}
+
+boolean useFixture = false
+String s3AccessKey = System.getenv("amazon_s3_access_key")
+String s3SecretKey = System.getenv("amazon_s3_secret_key")
+String s3Bucket = System.getenv("amazon_s3_bucket")
+String s3BasePath = System.getenv("amazon_s3_base_path")
+
+if (!s3AccessKey && !s3SecretKey && !s3Bucket && !s3BasePath) {
+  s3AccessKey = 'access_key'
+  s3SecretKey = 'secret_key'
+  s3Bucket = 'bucket'
+  s3BasePath = 'base_path'
+  useFixture = true
+
+} else if (!s3AccessKey || !s3SecretKey || !s3Bucket || !s3BasePath) {
+  throw new IllegalArgumentException("not all options specified to run against external S3 service are present")
+}
+
+if (useFixture) {
+  preProcessFixture {
+    dependsOn fixture.jar
+    doLast {
+      file("${testFixturesDir}/shared").mkdirs()
```

Comment:
It's not clear to me why all this is required. We seem to be effectively defining a new test fixture here. Why isn't we can't just use the existing `s3-fixture` like we do for other tests?

https://github.com/elastic/elasticsearch/blob/88ff69be5ddf185247826f62ca055e7727dcb521/plugins/repository-s3/build.gradle#L217

Reply:
This QA test is depends on ` repository-s3` project which already uses many S3 fixtures and the test fixtures plugin [prevented me from reusing](https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixtureExtension.java#L60-L62) the existing one. Since I wanted to avoid any mention to searchables snapshots in open source code I added the current configuration. Do you think there's a better way to do that?

Reply:
I would define a separate service in the [s3-fixture compose file](https://github.com/elastic/elasticsearch/blob/3e569014d180f969677dd287f8a57e2dbd4e36fb/test/fixtures/s3-fixture/docker-compose.yml) to be used here.

Reply:
> I would define a separate service in the s3-fixture compose file to be used here.

Sorry if I wasn't clear, I tried this solution and got a failure. Before doing what is coded in this PR I tried to define a separate service in the `/test/fixtures/s3-fixture/docker-compose.yml` file:

```yaml
...
  s3-fixture-generic:
    build:
      context: .
      args:
        fixtureClass: fixture.s3.S3HttpFixture
        port: 80
        bucket: "bucket"
        basePath: "base_path"
        accessKey: "access_key"
      dockerfile: ./testfixtures_shared/shared/Dockerfile
    volumes:
      - ./testfixtures_shared/shared:/fixture/shared
    ports:
      - "80"
```

and tried to declare the usage of this fixture in the `qa:s3` build file like this:
```java
if (useFixture) {
  testFixtures.useFixture(':test:fixtures:s3-fixture', 's3-fixture-generic')
}
```

But it failed with the following error: 
```
Build file '/x-pack/plugin/searchable-snapshots/qa/s3/build.gradle' line: 36

* What went wrong:
A problem occurred evaluating project ':x-pack:plugin:searchable-snapshots:qa:s3'.
> Projects :plugins:repository-s3 and :x-pack:plugin:searchable-snapshots:qa:s3 both claim all services from :test:fixtures:s3-fixture. This is not supported because it breaks running in parallel. Configure specific services in docker-compose.yml for each and add the service name to `useFixture`
```

In order to allow parallel test execution and to try to reuse as much as possible the existing code, I went with declaring a docker-compose.yml file in this `qa:s3`project that reuses the existing DockerFile and code from the test fixture.

Reply:
Understood. The problem here is that `:plugins:repository-s3` declares `testFixtures.useFixture(':test:fixtures:s3-fixture')` which means "give me _all_ services of the `s3-fixture` project. If we add a new service which that project doesn't need, we now need to be explicit say exactly which services it _does_ need. So you'll need to modify that build script and call `useFixture(':test:fixtures:s3-fixture', 's3-fixture')` and so on for each service required in the compose yaml.

I apologize for the back and forth, I just want to avoid us reimplementing a bunch of logic that our test fixtures plugin does for us if possible.

Reply:
>  The problem here is that `:plugins:repository-s3` declares `testFixtures.useFixture(':test:fixtures:s3-fixture')`

Thanks for spotting this @mark-vieira, I should have seen it by myself. I pushed 331408ad382e2dbebb50d4b7f7de204b54e5a913 to correct the fixtures.



> I apologize for the back and forth, I just want to avoid us reimplementing a bunch of logic that our test fixtures plugin does for us if possible.

Nothing to apologize, I pinged you because I was suspecting something wrong and you spotted it :+1:. I'm happy to not add another docker-compose.yml file to the mix :)